File: MergeSort.java
public class MergeSort
{

    public static void merge(int arr[],int start,int end)
    {
        int mid = start + end / 2;

        int flen = mid - start + 1;
        int slen = end - mid;

        int first[] = new int [flen];
        int second[] = new int[slen];

        int icnt =start;
        int jcnt = mid;
        int mainIndex = 0; 
        while(icnt < flen && jcnt < slen)
        {
            if(first[icnt] < second[jcnt])
            {
                arr[mainIndex++] = first[icnt++];
            }
            else
            {
                arr[mainIndex++] = second[jcnt++];
            }
        }

        while (icnt < flen) {
            arr[mainIndex++] = first[icnt++];
        }

        while(jcnt < slen)
        {
            arr[mainIndex++] = second[jcnt++];
        }
    }

    public static void mergesort(int arr[] , int start , int end)
    {
        if(start >= end)
        {
            return ;
        }
        int mid = start + end / 2;
        mergesort(arr, start, mid);
        mergesort(arr, mid+1, end);

        merge(arr,start,end);
    }
    public static void main(String args[])
    {
       int arr [] = new int[] {20,9,21,3};
       mergesort(arr,0,arr.length - 1);

       for(int i : arr)
       {
            System.out.println(i + " ");
       }
    }
}

File: MergeSort.cpp
class Solution{
    public:
    void merge(int arr[], int l, int m, int r)
    {
        int n1 = (m - l) + 1;
        int n2 = r - m;
        
        int L[n1];
        int R[n2];
        
        int k = l;
        
        for(int i = 0; i < n1 ; i++){
            L[i] = arr[k];
            k++;
        }
        
        for(int j = 0; j < n2 ; j++){
           R[j] = arr[k];
           k++;
        }
        
        k = l;
        int i = 0;
        int j = 0;
        
        while(i < n1 && j < n2){
            if(L[i] < R[j]){
                arr[k] = L[i];
                i++;
                k++;
            }
            else
            {
                arr[k] = R[j];
                k++;
                j++;
            }
        }
        
        
        while(i < n1){
             arr[k] = L[i];
                i++;
                k++;
        }
        
        while(j < n2){
           arr[k] = R[j];
                k++;
                j++;
        }
        
        
    }
public:
    void mergeSort(int arr[], int l, int r)
    {
        if(l >= r)
        {
            return;
        }
        
        int mid = l + (r-l) / 2;
        mergeSort(arr,l,mid);
        mergeSort(arr,mid+1,r);
        merge(arr,l,mid,r);
    }
};